arangod/Aql/Functions.cpp

```diff
@@ -6507,3 +6518,412 @@ AqlValue Functions::Fail(arangodb::aql::Query* query, transaction::Methods* trx,
+      return monthNames[static_cast<uint64_t>((unsigned)ymd.month()) - 1];
+    }},
+  {"%yyyy", [](std::string& wrk, tp_sys_clock_ms const& tp) {
+      auto ymd = year_month_day(floor<days>(tp));
+      auto yearnum = static_cast<int64_t>((int)year{ymd.year()});
+      if (yearnum < 0) {
+        if (yearnum > -10) {
+          return std::string("-000") + std::to_string(abs(yearnum));
+        }
+        if (yearnum > -100) {
+          return std::string("-00") + std::to_string(abs(yearnum));
+        }
+        if (yearnum > -1000) {
+          return std::string("-0") + std::to_string(abs(yearnum));
+        }
+        return std::string("-") + std::to_string(abs(yearnum));
+      }
+      if (yearnum < 0) {
+        return std::string("0000") + std::to_string(yearnum);
+      }
+
+      if (yearnum < 9) {
+        return std::string("000") + std::to_string(yearnum);
+      }
+      if (yearnum < 99) {
+        return std::string("00") + std::to_string(yearnum);
+      }
+      if (yearnum < 999) {
+        return std::string("0") + std::to_string(yearnum);
+      }
+
+      std::string yearstr(std::to_string(yearnum));
+      return tail(yearstr, 4);
+    }},
+
+  {"%wwww", [](std::string& wrk, tp_sys_clock_ms const& tp) {
+      weekday wd{floor<days>(tp)};
+      return weekDayNames[static_cast<uint64_t>(unsigned(wd))];
+    }},
+
+  {"%mmm", [](std::string& wrk, tp_sys_clock_ms const& tp) {
+      auto ymd = year_month_day(floor<days>(tp));
+      return monthNamesShort[static_cast<uint64_t>((unsigned)ymd.month()) - 1];
+    }},
+  {"%www", [](std::string& wrk, tp_sys_clock_ms const& tp) {
+      weekday wd{floor<days>(tp)};
+      return weekDayNamesShort[static_cast<uint64_t>(unsigned(wd))];
+    }},
+  {"%fff", [](std::string& wrk, tp_sys_clock_ms const& tp) {
+      auto day_time = make_time(tp - floor<days>(tp));
+      uint64_t millis = day_time.subseconds().count();
+      if (millis < 10) {
+        return std::string("00") + std::to_string(millis);
+      }
+      if (millis < 100) {
+        return std::string("0") + std::to_string(millis);
+      }
+      return std::to_string(millis);
+    }},
+  {"%xxx", [](std::string& wrk, tp_sys_clock_ms const& tp) {
+      auto ymd = year_month_day(floor<days>(tp));
+      auto yyyy = year{ymd.year()};
+      auto firstDayInYear = yyyy / 1 / 0;
+      uint64_t daysSinceFirst = duration_cast<days>(tp - sys_days(firstDayInYear)).count();
+      if (daysSinceFirst < 10) {
+        return std::string("00") + std::to_string(daysSinceFirst);
+      }
+      if (daysSinceFirst < 100) {
+        return std::string("0") + std::to_string(daysSinceFirst);
+      }
+      return std::to_string(daysSinceFirst);
+    }},
+  
+  // there"s no really sensible way to handle negative years, but better not drop the sign
+  {"%yy", [](std::string& wrk, tp_sys_clock_ms const& tp) {
+      auto ymd = year_month_day(floor<days>(tp));
+      auto yearnum = static_cast<int64_t>((int)year{ymd.year()});
```

Comment:
why casting to "int" here?

Reply:
CONSTCD11 explicit operator int() const NOEXCEPT;
is used here.